beacon_node/beacon_chain/src/block_status_table.rs

```diff
@@ -0,0 +1,784 @@
+    pub fn is_terminal(&self) -> bool {
+        matches!(self, BlockStatus::Invalid | BlockStatus::Imported)
+    }
+}
+
+/// A thread-safe status tracking table for Ethereum blocks during data availability import.
+///
+/// This table uses DashMap for lock-free concurrent access and AtomicU8 values for
+/// efficient atomic operations on status transitions to prevent race conditions during
+/// block processing.
+#[derive(Debug)]
+pub struct BlockStatusTable {
+    /// The underlying concurrent hash map storing block status as AtomicU8
+    table: DashMap<Hash256, AtomicU8>,
+    /// Broadcast channels for notifying subscribers of status changes
+    notifiers: DashMap<Hash256, broadcast::Sender<BlockStatus>>,
+    /// Counter for tracking the number of entries (for metrics/debugging)
+    entry_count: AtomicUsize,
+}
+
+impl BlockStatusTable {
+    /// Channel capacity - sufficient for all possible state transitions
+    const CHANNEL_CAPACITY: usize = 16;
+
+    /// Creates a new empty status table.
+    pub fn new() -> Self {
+        Self {
+            table: DashMap::new(),
+            notifiers: DashMap::new(),
+            entry_count: AtomicUsize::new(0),
+        }
+    }
+
+    /// Subscribe to status updates for a specific block.
+    /// Returns None if the block is not being tracked.
+    ///
+    /// Late subscribers will receive all future status changes but not past ones.
+    pub fn subscribe(
+        &self,
+        block_hash: &Hash256,
+    ) -> Option<(BlockStatus, broadcast::Receiver<BlockStatus>)> {
+        // by subscribing before reading the current status, a race condition is possible where
+        // the subscribing thread receives a status and then when they await the channel, they receive
+        // the same status again. This is preferable to the reverse race condition where an old status is
+        // read and the subscriber doesn't send the update to the latest status.
+        let rx = self.notifiers.get(block_hash)?.subscribe();
+        let current_status = self.get_status(block_hash)?;
+        Some((current_status, rx))
+    }
+
+    /// Gets the current status of a block.
+    ///
+    /// Returns `None` if the block is not tracked in the table.
+    pub fn get_status(&self, block_hash: &Hash256) -> Option<BlockStatus> {
+        self.table.get(block_hash).and_then(|entry| {
+            let value = entry.value().load(Ordering::Acquire);
+            BlockStatus::from_u8(value)
+        })
+    }
+
+    /// Atomically compares and swaps the status of a block.
+    ///
+    /// This operation is atomic and will only succeed if the current status
+    /// matches the expected status. Returns `true` if the swap succeeded,
+    /// `false` otherwise.
+    ///
+    /// # Arguments
+    /// * `block_hash` - The hash of the block to update
+    /// * `expected` - The expected current status
+    /// * `new` - The new status to set
+    pub fn compare_and_swap_status(
```

Comment:
This function returns `false` for two very distinct cases:
- if the block is not known
- if the expect status value doesn't match

I imagine in practice we just get a mutable reference to the dashmap and match over it. For example in sync

```rust
match block_status_map.get_mut(block_hash) {
    Pending => todo!?(),
    Executing(handle) => handle.await,
    Invalid => reject_block(),
    Importing(handle) => handle.await,
    Imported => complete_lookup_and_continue(),
}
```

Reply:
I imagine in practice we just get a mutable reference to the dashmap and match over it. For example in sync

```rust
match block_status_map.get_mut(block_hash) {
    Pending => todo!?(),
    Executing(handle) => handle.await,
    Invalid => reject_block(),
    Importing(handle) => handle.await,
    Imported => complete_lookup_and_continue(),
}
```

Reply:
I've changed this function to return an error that will tell exactly what happened. Thanks! :D